addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/filesystem-role-service.js

```diff
@@ -0,0 +1,553 @@
+    // - For each found fs role arn, load its fs role entity,
+    //   - Does the entity include the member account in its trust and include the study with the same envPermission,
+    //     if so, then we don't need to create an fs role
+    //   - If not, can we fit the member account in the trust policy, if so then update the actual trust document
+    //     in the data source account for the fs role, then update the database, see item #3 above.
+    // - If no fs role entity is found, or the existing on has its trust property full, we need to create a new
+    //   fs role as follows:
+    //   - Create the fs role entity and add the member account to the trust property, see item #3 above.
+    //     - At this point, we don't store the entity in the data base until after we are able to create
+    //       it in the data source account
+    //   - Assume the app role and create the fs role in the data source account
+    //   - Store the entity in the database
+    //   - Add usage of the study by this role, see item #1 above.
+
+    // Get app role entity
+    const appRoleEntity = await appRoleService.mustFind(requestContext, { accountId, bucket, arn: appRoleArn });
+
+    // Get all existing usages by fs roles for this study (keeping env permission in mind)
+    const fsRoleUsages = await resourceUsageService.getResourceUsage(requestContext, {
+      resource: studyResourceId,
+      setName: fsUsageSetName,
+    });
+    const fsRoleEntities = _.get(fsRoleUsages, fsUsageSetName, []);
+
+    // Start a loop to see if existing fs roles for the study have the member account already
+    // or can take it without exceeding the trust doc limit?
+    let fsRoleEntity;
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const fsRoleArn of fsRoleEntities) {
+      const entity = await this.find(requestContext, { accountId, bucket, arn: fsRoleArn });
+      // eslint-disable-next-line no-continue
+      if (!entity) continue;
+
+      // Check if the provided study matches any of the studies in the fs role entity, including the env permissions
+      // eslint-disable-next-line no-continue
+      if (!hasStudy(entity, studyEntity)) continue;
+
+      if (hasMemberAccount(entity, memberAccountId)) {
+        // We found an existing fs role entity that already has the member account
+        fsRoleEntity = entity;
+
+        // Add environment usage
+        await resourceUsageService.addUsage(requestContext, {
+          resource: studyResourceId,
+          setName: memberAccountId,
+          item: envId,
+        });
+
+        break;
+      }
+
+      if (!maxReached(addMemberAccount(entity, memberAccountId))) {
+        // We found an existing fs role entity and we can fit the member account in its trust doc
+
+        // Update the fs role policy in the data source account
+        await this.updateAssumeRolePolicy(entity);
```

Comment:
Does this take care of duplicate member account additions? Basically we shouldn't add duplicates.

Reply:
Yes. The entity has an array named 'trust' and since use the 'addMemberAccount' to add accounts, the 'addMemberAccount' won't add an account to the 'trust' array if it is already there.